### PR TITLE
Ioperation Unit tests for IForLoopStatment

### PR DESCRIPTION
**Pull request by jinujoseph**
Part of https://github.com/dotnet/roslyn/issues/17602 
cc @dotnet/analyzer-ioperation , @cston

### CODE REVIEW

**Comment by jinujoseph** [comment]
test windows_debug_unit32_prtest

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.vb [hunk 1]
```diff
@@ -0,0 +1,896 @@
...
+        For AVarName = 1 To 2'BIND:"For AVarName = 1 To 2"
+            For B = 1 To 2
+                For C = 1 To 2
+                    For D = 1 To 2
+                    Next D
+                Next C
+            Next B
+        Next AVarName
+    End Sub
+End Class
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.vb]: Consider testing combined `Next`:
```
For A = 1 To 2
    For B = A To 2
Next B, A
```

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.vb [hunk 2]
```diff
@@ -0,0 +1,896 @@
...
+  AtLoopBottom:
+      IExpressionStatement (OperationKind.ExpressionStatement) (Syntax: 'For i As In ... Next')
+        Expression: ICompoundAssignmentExpression (BinaryOperationKind.IntegerAdd) (OperationKind.CompoundAssignmentExpression, Type: System.Int32) (Syntax: 'For i As In ... Next')
+            Left: ILocalReferenceExpression: i (OperationKind.LocalReferenceExpression, Type: System.Int32) (Syntax: 'i As Integer')
+            Right: IConversionExpression (ConversionKind.Cast, Explicit) (OperationKind.ConversionExpression, Type: System.Int32, Constant: 1) (Syntax: 'For i As In ... Next')
+                Operand: ILiteralExpression (OperationKind.LiteralExpression, Type: System.Int32, Constant: 1) (Syntax: 'For i As In ... Next')
+  Body: IBlockStatement (1 statements) (OperationKind.BlockStatement) (Syntax: 'For i As In ... Next')
+      IIfStatement (OperationKind.IfStatement) (Syntax: 'If i Mod 2  ... End If')
+        Condition: IBinaryOperatorExpression (BinaryOperationKind.IntegerEquals) (OperationKind.BinaryOperatorExpression, Type: System.Boolean) (Syntax: 'i Mod 2 = 0')
+            Left: IBinaryOperatorExpression
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.vb]: Should `'BIND:` refer to `For X = 1 To 10` instead? Can we remove this method?

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.vb (hunk 1 above)
Comment (jinujoseph) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.vb]: oh interesting , will do

Comment (jinujoseph) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.vb]: done

**Comment by cston** [comment]
LGTM

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.vb [hunk 3]
```diff
@@ -0,0 +1,996 @@
...
+            System.Console.WriteLine(myarray(i))
+        Next
+    End Sub
+End Class
+    ]]>.Value
+
+            Dim expectedOperationTree = <![CDATA[
+IForLoopStatement (LoopKind.For) (OperationKind.LoopStatement) (Syntax: 'For i As In ... Next')
+  Condition: IBinaryOperatorExpression (BinaryOperationKind.IntegerLessThanOrEqual) (OperationKind.BinaryOperatorExpression, Type: System.Boolean) (Syntax: 'myarray.Length - 1')
+      Left: ILocalReferenceExpression: i (OperationKind.LocalReferenceExpression, Type: System.Int32) (Syntax: 'i As Integer')
```
Comment (heejaechang) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.vb]: @dotnet/analyzer-ioperation @mavasani I noticed that in vb, ILocationReferenceExpresssion here are reused in other place as well making Operation not a tree but graph.

I think we need to modify our Operation Unit test to fail if same operation node is used multiple time in the tree.

Comment (jinujoseph) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.vb]: I agree , Tracking https://github.com/dotnet/roslyn/issues/21068 to fix this

**Pull request merged**